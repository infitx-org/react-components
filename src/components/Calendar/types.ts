export type PossibleDate = Date | undefined;

export type PossibleDay = string | undefined;

export type Matrix = PossibleDay[][];

export type DisabledDays = (day: Date) => boolean;

export type DateRange = [PossibleDate, PossibleDate];
