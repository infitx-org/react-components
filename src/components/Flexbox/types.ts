export enum TopBottom {
  "top" = "flex-start",
  "bottom" = "flex-end",
}
export enum LeftRight {
  "left" = "flex-start",
  "right" = "flex-end",
}

export enum Mappers {
  "top" = "flex-start",
  "bottom" = "flex-end",
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap<T extends string> = `${AlignItems}` | `${T}`;
export type JustifyWithMap<T extends string> = `${JustifyContent}` | `${T}`;
