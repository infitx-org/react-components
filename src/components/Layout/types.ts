export enum AlignMap {
  "top" = "flex-start",
  "bottom" = "flex-end",
}

export enum JustifyMap {
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap = `${AlignItems | keyof typeof AlignMap}`;
export type JustifyWithMap = `${JustifyContent | keyof typeof JustifyMap}`;
