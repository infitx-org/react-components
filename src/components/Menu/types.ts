export interface MenuSectionProps {
  label: string;
  hidden?: boolean;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  size?: number;
  disabled?: boolean;
  children: MenuItemElement | MenuItemElement[];
}

interface BaseMenuItemProps {
  label: string;
  active?: boolean;
  disabled?: boolean;
  hidden?: boolean;
  back?: boolean;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  size?: number;
  // prop partial is not used directly by the MenuItem but it is used by the Menu component
  partial?: boolean;
  children?: MenuItemElement | MenuElement[];
}

interface BaseMenuItemPropsWithTo extends BaseMenuItemProps {
  to: string;
  path?: string;
}
interface BaseMenuItemPropsWithPath extends BaseMenuItemProps {
  to?: string;
  path: string;
}

export type MenuItemProps = BaseMenuItemPropsWithTo | BaseMenuItemPropsWithPath;

export type MenuItemElement = React.ReactElement<MenuItemProps>;
export type MenuSectionElement = React.ReactElement<MenuSectionProps>;
export type MenuElement = MenuItemElement | MenuSectionElement;
