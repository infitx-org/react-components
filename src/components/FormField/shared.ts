import { ButtonProps as RawButtonProps } from "../Button";
import { CheckboxProps as RawCheckboxProps } from "../Checkbox/Checkbox";
import { DatePickerProps as RawDatePickerProps } from "../DatePicker/DatePicker";
import { FileUploaderProps as RawFileUploaderProps } from "../FileUploader/FileUploader";
import { NumberFieldProps as RawNumberFieldProps } from "../NumberField/NumberField";
import { RadioGroupProps as RawRadioGroupProps } from "../RadioGroup/RadioGroup";
import { SelectProps as RawSelectProps } from "../Select/Select";
import { TextFieldProps as RawTextFieldProps } from "../TextField/TextField";
import "./FormField.scss";

interface ButtonFormProps extends RawButtonProps {
  type: "button";
  label: string;
}
interface CheckboxFormProps extends RawCheckboxProps {
  type: "checkbox";
  label: string;
}
interface DatePickerFormProps extends RawDatePickerProps {
  type: "date";
  label: string;
}
interface FileUploaderFormProps extends RawFileUploaderProps {
  type: "file";
  label: string;
}
interface NumberFieldFormProps extends RawNumberFieldProps {
  type: "number";
  label: string;
}
interface RadioGroupFormProps extends RawRadioGroupProps {
  type: "radio";
  label: string;
}
interface SelectFormProps extends RawSelectProps {
  type: "select";
  label?: string;
}
interface TextFieldFormProps extends RawTextFieldProps {
  type: "text" | "password";
  label?: string;
}

export type BaseFieldProps =
  | ButtonFormProps
  | CheckboxFormProps
  | DatePickerFormProps
  | FileUploaderFormProps
  | NumberFieldFormProps
  | RadioGroupFormProps
  | SelectFormProps
  | TextFieldFormProps;

export function isButton(props: BaseFieldProps): props is ButtonFormProps {
  return props.type === "button";
}

export function isTextField(
  props: BaseFieldProps
): props is TextFieldFormProps {
  return props.type === "text" || props.type === "password";
}

export function isDatePicker(
  props: BaseFieldProps
): props is DatePickerFormProps {
  return props.type === "date";
}

export function isFileUploader(
  props: BaseFieldProps
): props is FileUploaderFormProps {
  return props.type === "file";
}

export function isNumberField(
  props: BaseFieldProps
): props is NumberFieldFormProps {
  return props.type === "number";
}

export function isSelect(props: BaseFieldProps): props is SelectFormProps {
  return props.type === "select";
}

export function isRadio(props: BaseFieldProps): props is RadioGroupFormProps {
  return props.type === "radio";
}

export function isCheckbox(props: BaseFieldProps): props is CheckboxFormProps {
  return props.type === "checkbox";
}

export type FormFieldProps = BaseFieldProps & {
  outerDirection?: "row" | "column";
};
