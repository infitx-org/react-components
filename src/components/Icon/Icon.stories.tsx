import React from "react";
import { Size } from "types";
import Icon from "./Icon";

export default {
  title: "Icon",
  component: Icon,
};

const icon = (
  <svg viewBox="0 0 100 100">
    <circle cx="50" cy="20" r="10" />
    <circle cx="50" cy="80" r="10" />
    <circle cx="20" cy="50" r="10" />
    <circle cx="80" cy="50" r="10" />
  </svg>
);

const Template = (args) => <Icon icon={icon} size={50} {...args} />;

export const FillColor = Template.bind({});
FillColor.args = {
  fill: "#f00",
};

export const StrokeColor = Template.bind({});
StrokeColor.args = {
  stroke: "#f00",
  fill: "transparent",
  size: 50,
};

export const SizeLarge = Template.bind({});
SizeLarge.args = {
  size: Size.Large,
};

export const SizeMedium = Template.bind({});
SizeMedium.args = {
  size: Size.Medium,
};

export const SizeSmall = Template.bind({});
SizeSmall.args = {
  size: Size.Small,
};

export const SizeXSmall = Template.bind({});
SizeXSmall.args = {
  size: Size.XSmall,
};
