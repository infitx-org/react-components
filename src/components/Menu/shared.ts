
import React from "react";
import MenuItem from './components/MenuItem';
import MenuSection from './components/MenuSection';

export interface MenuSectionProps {
  label: string;
  hidden?: boolean;
  icon?: string;
  fill?: string;
  size?: number;
  disabled?: boolean;
  children: MenuItemElement[];
}

export interface MenuItemProps {
  label: string;
  active: boolean;
  path?: string;
  to?: string;
  disabled?: boolean;
  hidden?: boolean;
  back?: boolean;
  icon?: string;
  fill?: string;
  size?: number;
  // prop asRoot is not used directly by the MenuItem but it is used by the Menu component
  // eslint-disable-next-line react/no-unused-prop-types
  asRoot?: boolean;
  // prop partial is not used directly by the MenuItem but it is used by the Menu component
  // eslint-disable-next-line react/no-unused-prop-types
  partial?: boolean;
  // eslint-disable-next-line react/no-unused-prop-types
  children?: MenuElement[];
}

export type MenuItemElement = React.ReactElement<MenuItemProps>;
export type MenuSectionElement = React.ReactElement<MenuSectionProps>;
export type MenuElement = MenuItemElement | MenuSectionElement;

export interface MenuContextValue {
  pathname: string;
  onClick: (path: string) => void;
}

export const MenuContext = React.createContext<MenuContextValue>({
  pathname: "",
  onClick: () => undefined,
});

export function getPathMatches(
  pathname: string,
  path: string | undefined,
  partial: boolean | undefined
): boolean {
  if (!path) {
    return false;
  }
  if (partial && pathname.startsWith(path)) {
    return true;
  }
  if (path === pathname) {
    return true;
  }

  const pathChunks = path.split("/");
  const pathNameChunks = pathname.split("/");

  if (pathChunks.length !== pathNameChunks.length) {
    return false;
  }

  return pathNameChunks.every((chunk, index) => {
    const pathChunk = pathChunks[index];
    const isExact = pathChunk === chunk;
    const isWild = pathChunk !== undefined && pathChunk.startsWith(":");
    return isExact || isWild;
  });
}

export function isMenuItem(child: React.ReactNode): child is MenuItemElement {
  return (child as React.ReactElement).type === MenuItem;
}

export function isMenuSection(child: React.ReactNode): child is MenuSectionElement {
  return (child as React.ReactElement).type === MenuSection;
}