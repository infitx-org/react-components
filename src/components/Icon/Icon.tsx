import React from "react";
import { Size } from "../types";
import { getIconSizeByComponentSize } from "../shared";

export type IconProps = {
  icon: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  stroke?: string;
  size?: `${Size}` | number;
  id?: string;
  style?: React.CSSProperties;
  className?: string;
};

function override(
  fallback: string,
  intrinsic?: string,
  defined?: string
): string {
  return defined || intrinsic || fallback;
}

function Icon({
  icon,
  fill,
  stroke,
  size = 20,
  id,
  style,
  className,
}: IconProps) {
  return React.cloneElement(icon, {
    ...icon.props,
    fill: override("#000", icon.props.fill, fill),
    stroke: override("transparent", icon.props.stroke, stroke),
    width: typeof size === "string" ? getIconSizeByComponentSize(size) : size,
    height: typeof size === "string" ? getIconSizeByComponentSize(size) : size,
    id,
    style,
    className,
  });
}

export default React.memo(Icon);
