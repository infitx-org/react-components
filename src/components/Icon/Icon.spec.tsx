import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Kind, Size } from "types";
import Icon from "./Icon";
import { getIconSizeByComponentSize } from "../shared";

const icon = (
  <svg width="55" height="55" viewBox="-6 -3 55 55">
    <circle cx="10" cy="10" r="20" />
  </svg>
);

describe("tests the icon", () => {
  it("renders the icon", () => {
    const { container } = render(<Icon icon={icon} />);
    expect(container.querySelector("svg")).toBeTruthy();
    expect(container.querySelector("circle")).toBeTruthy();
  });

  it("renders the fill prop", () => {
    const { container } = render(<Icon icon={icon} fill="#f00" />);
    expect(container.querySelector("svg")).toHaveAttribute("fill", "#f00");
  });

  it("renders the stroke prop", () => {
    const { container } = render(<Icon icon={icon} stroke="#f00" />);
    expect(container.querySelector("svg")).toHaveAttribute("stroke", "#f00");
  });

  it("renders the numeric size prop", () => {
    const { container } = render(<Icon icon={icon} size={100} />);
    expect(container.querySelector("svg")).toHaveAttribute("height", "100");
    expect(container.querySelector("svg")).toHaveAttribute("width", "100");
  });

  it("renders the string size prop", () => {
    Object.values(Size).forEach((size) => {
      const { container } = render(<Icon icon={icon} size={size} />);
      expect(container.querySelector("svg")).toHaveAttribute(
        "height",
        getIconSizeByComponentSize(size).toString()
      );
      expect(container.querySelector("svg")).toHaveAttribute(
        "width",
        getIconSizeByComponentSize(size).toString()
      );
    });
  });

  it("renders the id prop", () => {
    const { container } = render(<Icon id="test" icon={icon} />);
    expect(container.querySelector("svg#test")).toBeTruthy();
  });

  it("renders the className prop", () => {
    const { container } = render(<Icon className="test" icon={icon} />);
    expect(container.querySelector("svg.test")).toBeTruthy();
  });

  it("renders the style prop", () => {
    const { container } = render(<Icon style={{ fill: "red" }} icon={icon} />);
    expect(container.querySelector("svg")).toHaveAttribute(
      "style",
      "fill: red;"
    );
  });

  it("renders the default fill prop", () => {
    const { container } = render(<Icon icon={icon} />);
    expect(container.querySelector("svg")).toHaveAttribute("fill", "#000");
  });

  it("renders the default stroke prop", () => {
    const { container } = render(<Icon icon={icon} />);
    expect(container.querySelector("svg")).toHaveAttribute(
      "stroke",
      "transparent"
    );
  });

  it("renders the default numeric size prop", () => {
    const { container } = render(<Icon icon={icon} />);
    expect(container.querySelector("svg")).toHaveAttribute("height", "20");
    expect(container.querySelector("svg")).toHaveAttribute("width", "20");
  });
});

// Snapshot testing
it("renders the Icon correctly when multiple props are set", () => {
  const { container } = render(<Icon icon={icon} fill="#f00" stroke="#000" />);
  expect(container).toMatchSnapshot();
});
