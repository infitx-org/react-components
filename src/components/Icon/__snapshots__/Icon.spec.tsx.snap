// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Icon correctly when multiple props are set 1`] = `
<div>
  <svg
    fill="#f00"
    height="20"
    stroke="#000"
    viewBox="-6 -3 55 55"
    width="20"
  >
    <circle
      cx="10"
      cy="10"
      r="20"
    />
  </svg>
</div>
`;
